### PR TITLE
Fail if a remote source content doesn't match lockfile

### PR DESCRIPTION
**Pull request by stevendanna**
If a URL based source does not match the shasum recorded in the
lockfile, it likely means a new version has been pushed to the remote
source. In this case, we fail to help ensure that when using a lockfile
we always run the same code as when the lockfile was created.

Signed-off-by: Steven Danna steve@chef.io

### CODE REVIEW

File: lib/fetchers/url.rb [hunk 1]
```diff
@@ -81,34 +82,61 @@ def initialize(url, opts)
     end
 
     def fetch(path)
-      Inspec::Log.debug("Fetching URL: #{@target}")
-      @archive_path = download_archive(path)
+      @archive_path ||= download_archive(path)
+    end
+
+    def shasum
```
Comment (chris-rock) [inline lib/fetchers/url.rb]: maybe simply `sha256`?

File: lib/fetchers/local.rb [hunk 2]
```diff
@@ -55,8 +55,19 @@ def archive_path
       @target
     end
 
+    def cache_key
+      archive_shasum.to_s
+    end
+
+    def archive_shasum
```
Comment (chris-rock) [inline lib/fetchers/local.rb]: maybe just `sha256`?

File: lib/fetchers/url.rb [hunk 3]
```diff
@@ -81,34 +82,61 @@ def initialize(url, opts)
...
+                  File.read(@archive_path)
+                else
+                  remote_archive_content
+                end
+      Digest::SHA256.hexdigest content
     end
 
     def resolved_source
-      { url: @target }
+      @resolved_source ||= { url: @target, shasum256: shasum }
```
Comment (chris-rock) [inline lib/fetchers/url.rb]: `shasum256` What do you think about `sha256`?

File: lib/fetchers/url.rb [hunk 4]
```diff
@@ -81,34 +82,61 @@ def initialize(url, opts)
...
+    def open_target
+      Inspec::Log.debug("Fetching URL: #{@target}")
       http_opts = {}
       http_opts['ssl_verify_mode'.to_sym] = OpenSSL::SSL::VERIFY_NONE if @insecure
       http_opts['Authorization'] = "Bearer #{@token}" if @token
+      open(@target, http_opts)
+    end
 
-      remote = open(@target, http_opts)
+    def remote_archive_content
```
Comment (chris-rock) [inline lib/fetchers/url.rb]: maybe just `content`?

**Comment by stevendanna** [comment]
Suggested renames pushed

File: lib/inspec/fetcher.rb [hunk 5]
```diff
@@ -16,7 +16,7 @@ def resolve(target)
       end
     end
 
-    NON_FETCHER_KEYS = [:name, :version_constraint, :cwd, :backend, :cache].freeze
+    NON_FETCHER_KEYS = [:name, :version_constraint, :cwd, :backend, :cache, :shasum256].freeze
```
Comment (chris-rock) [inline lib/inspec/fetcher.rb]: should this be `sha256`?

**Comment by chris-rock** [comment]
fixes #1066

**Comment by chris-rock** [comment]
Thanks @stevendanna

**Pull request merged**